src/kademlia/dht_storage.cpp

```diff
@@ -0,0 +1,355 @@
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+/*
+
+Second author: Alden Torres (aldenml)
+Same copyright
+
+*/
+
+#include "libtorrent/kademlia/dht_storage.hpp"
+
+#include <utility>
+#include <boost/bind.hpp>
+#include <boost/function/function1.hpp>
+#include <boost/tuple/tuple.hpp>
+
+#include <libtorrent/socket_io.hpp>
+#include <libtorrent/aux_/time.hpp>
+
+namespace libtorrent { namespace dht
+{
+	namespace
+	{
+#ifndef TORRENT_NO_DEPRECATE
+		struct count_peers
+		{
+			int* count;
+			count_peers(int* c): count(c) {}
+			void operator()(std::pair<libtorrent::sha1_hash
+					, libtorrent::dht::torrent_entry> const& t)
+			{
+				*count += t.second.peers.size();
+			}
+		};
+#endif
+
+		// remove peers that have timed out
+		void purge_peers(std::set<peer_entry>& peers)
+		{
+			// TODO: 2 make this configurable in dht_settings
+			int announce_interval = 30;
+
+			for (std::set<peer_entry>::iterator i = peers.begin()
+				 , end(peers.end()); i != end;)
+			{
+				// the peer has timed out
+				if (i->added + minutes(int(announce_interval * 1.5f)) < aux::time_now())
+					peers.erase(i++);
+				else
+					++i;
+			}
+		}
+
+		void touch_item(dht_popularity_item* f, address const& address)
+		{
+			f->last_seen = aux::time_now();
+
+			// maybe increase num_announcers if we haven't seen this IP before
+			sha1_hash iphash;
+			hash_address(address, iphash);
+			if (!f->ips.find(iphash))
+			{
+				f->ips.set(iphash);
+				++f->num_announcers;
+			}
+		}
+
+		// return true of the first argument is a better candidate for removal, i.e.
+		// less important to keep
+		struct immutable_item_comparator
+		{
+			immutable_item_comparator(sha1_hash const& our_id) : m_our_id(our_id) {}
+			immutable_item_comparator(immutable_item_comparator const& c)
+				: m_our_id(c.m_our_id) {}
+
+			bool operator() (std::pair<sha1_hash, dht_immutable_entry> const& lhs
+				, std::pair<sha1_hash, dht_immutable_entry> const& rhs) const
+			{
+				int l_distance = distance_exp(lhs.first, m_our_id);
+				int r_distance = distance_exp(rhs.first, m_our_id);
+
+				// this is a score taking the popularity (number of announcers) and the
+				// fit, in terms of distance from ideal storing node, into account.
+				// each additional 5 announcers is worth one extra bit in the distance.
+				// that is, an item with 10 announcers is allowed to be twice as far
+				// from another item with 5 announcers, from our node ID. Twice as far
+				// because it gets one more bit.
+				return lhs.second.popularity.num_announcers / 5 - l_distance <
+					rhs.second.popularity.num_announcers / 5 - r_distance;
+			}
+
+		private:
+
+			// explicitly disallow assignment, to silence msvc warning
+			immutable_item_comparator& operator=(immutable_item_comparator const&);
+
+			node_id const& m_our_id;
+		};
+	}
+
+	// internal
+	inline bool operator<(ed25519_public_key const& lhs, ed25519_public_key const& rhs)
```

Comment:
I don't think these have to be inline anymore. also, they should probably be in the anonymous namespace